### ISSUE TITLE
Customize the default key naming logic/process in generated json from getters

### ISSUE THREAD

**Issue by dhruv-niki** [stleary/JSON-java#401]
Hi,
It seems the default logic to generate the json keys based on public getters is something like this:
1. Find public functions named: get[A-Z]+
2. In generated json, the key name is the part after 'get' with the first letter in lowercase - but case 3 below indicates that's not it.

The auto naming works great most of the time, but I have one particular case where it's an issue.
I need the json output to be: `{ "mCode": "xyz" }`

If my getter is named --> json generated is:
1. `getmCode()` --> not picked up at all, since lowercase letter after 'get'
2. `getmcode()` --> not picked up at all, since lowercase letter after 'get'
3. `getMCode()` --> `"MCode": "xyz"`
4. `getMcode()` --> `"mcode": "xyz"`
5. `getMcOde()` --> `"mcOde": "xyz"`

So I can see no way of automatically getting "mCode": "xyz" in the json. For now I am manually adding it to the JSONObject.

Can there be a way to custom define the key naming logic for a particular getter, or an entire class?

**Comment by johnjaylward** [comment]
@stleary the issue seems to be here:
https://github.com/stleary/JSON-java/blob/master/JSONObject.java#L1449
```java
if (key.length() > 0
      && Character.isUpperCase(key.charAt(0))) {
```

I'm not sure why the uppercase check is in there. Looks like it's been there at least 5 years and was put in by Douglas.

It does seem like something we shouldn't be selecting on. It looks like if we remove the uppercase check there, it should not make the autonaming logic change, just be more inclusive.

If removed, @dhruv-niki 's examples would look more like this:
1. `getmCode()` --> `"mCode": "xyz"`
2. `getmcode()` --> `"mcode": "xyz"`
3. `getMCode()` --> `"MCode": "xyz"`
4. `getMcode()` --> `"mcode": "xyz"`
5. `getMcOde()` --> `"mcOde": "xyz"`

**Comment by stleary** [comment]
I would guess that the original intent was to support Netbean getters. Not convinced it is it should be changed. Users who need this functionality should probably just fork the project. However, OK to go ahead if anyone:
* Feels strongly about supporting this
* Can write the code in a way that does not affect existing applications
* Can update the unit tests as needed

**Comment by mouse07410** [comment]
> I'm not sure why the uppercase check is in there. Looks like it's been there at least 5 years and was put in by Douglas.

Not meaning to insult, but it looks like an example of shortsightedness in design.

> Not convinced it should be changed.

I am 100% convinced it should be changed, and suggest removing this pointless restriction.

It doesn't make sense to create a fork merely to remedy such an obvious oversight - unless the maintainers are too closely attached to this design mistake.

As you can see from my tone, I feel **very strongly** about this. ;-)
But probably won't write the code, though it's highly unlikely that a straightforward removal of this uppercase conversion would affect any sane code.

**Comment by stleary** [comment]
> Not meaning to insult, but it looks like an example of shortsightedness in design.

No worries, feel free to bring this up with the library author. You may also want to mention any examples of poor design that you have found in JSON itself, _since he invented it_.

**Comment by strkkk** [comment]
> Can write the code in a way that does not affect existing applications

Since it uses reflection for generation name of the property, maybe some annotation can be added, which defines exact name of the generated property, e.g.
```
@JsonPropertyName("mCode") 
public String getmCode() {return someField;}
```
And if such annotation exists, its value is taken as property name.
This will not break existing code.

**Comment by erosb** [comment]
+1 for `@JSONPropertyIgnore` , I find it a useful feature.

**Comment by stleary** [comment]
I believe this issue has been addressed in #406. If you think otherwise, let me know.
